compiler/rustc_middle/src/ty/predicate.rs

```diff
@@ -290,7 +290,7 @@ impl<'tcx> ty::List<ty::PolyExistentialPredicate<'tcx>> {
     /// have a "trivial" vtable consisting of just the size, alignment,
     /// and destructor.
     pub fn principal(&self) -> Option<ty::Binder<'tcx, ExistentialTraitRef<'tcx>>> {
-        self[0]
+        self.get(0)?
```

Comment:
I don't like changing this invariant in `rustc_middle` just because CFI makes bad types 😞

Reply:
No other place in the compiler expects or constructs `dyn` types with no predicates.

Reply:
Since later on in the type encoding process, this type will be examined under normalization, I can't just guard against calling `principal` in the calling code. Given your comment, that probably means I need to avoid constructing 0-predicate `dyn`

Do you have suggestions for predicates or representations that would make sense here? The case in question is that I have a `dyn Send` and `dyn Sync` - they share a vtable. This vtable has one method, `drop_in_place`. For purposes of checking the alias set, do you have opinions on what it should be encoded as? The only thing I will know about this object at the method site is that it is a `dyn` with no principal.

Three possible encodings that come to mind:
* `dyn Drop` - I don't like this because these types may not implement `Drop`, and methods which require `fn drop` shouldn't go here, but it is descriptive.
* `dyn Sized` - I'm considering this specifically because you *can't* normally have a `dyn Sized` object, and so it can't accidentally match anything else.
* Contract the *entire* receiver to `*mut ()` - this technically discards the alias set restriction that someone made a `dyn` out of this type at some point in the program, but should otherwise be fine.

I'm currently adjusting my code to use the `*mut ()` approach, but figured I'd lay out my thoughts and see if you have a different opinion.

Reply:
You could perhaps use [`dyn Destruct`](https://doc.rust-lang.org/std/marker/trait.Destruct.html), which every type currently implements (and approximately matches the fact that all we know is that it has a `drop_in_place`)?

`*mut ()` also works I guess.